File: doot_example/c/doot.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct Node {
    char letter;
    struct Node* next;
    struct Node* last;
};

void printList(struct Node* n){
    printf("Printing List!\n");
    while(n!= NULL){
        printf("%c",n->letter);
        n=n->next;
    }
    printf("\n");
}

void makeList(char *c, struct Node* head){
    printf("Received: %s\n", c);
    int i = 0;
    while (c[i] != '\0'){
        printf("Loading %c into node %d\n", c[i], i);
        //if empty, 
        if(head->letter==NULL){
            head->letter = c[i];
        //else second entry
        }else if(head->next==NULL){
            struct Node* node = malloc(sizeof(struct Node)); 
            node->letter = c[i];
            node->next = NULL;
            node->last = NULL;
            head->next = node;
            head->last = node;
           //printf("Sanity Checking: head->letter=%c\n", head->letter);
        //else nth entry
        }else {
            //printf("Sanity Checking: head->letter=%c\n", head->letter);
            struct Node* node = malloc(sizeof(struct Node)); 
            node->letter = c[i];
            node->next = NULL;
            node->last = NULL;
            head->last->next = node;
            head->last = node;
        }
        i++;
    }

}

int main(){
    
    struct Node* head = NULL; 
    struct Node* second = NULL;
    struct Node* third = NULL;
    struct Node* fourth = NULL;
    struct Node* fifth = NULL;

    head = (struct Node*)malloc(sizeof(struct Node));
    second = (struct Node*)malloc(sizeof(struct Node));
    third = (struct Node*)malloc(sizeof(struct Node));
    fourth = (struct Node*)malloc(sizeof(struct Node));
    fifth = (struct Node*)malloc(sizeof(struct Node));

    head->letter = 'D';
    head->next = second;

    second->letter = 'o';
    second->next = third;

    third->letter = 'o';
    third->next = fourth;

    fourth->letter = 't';
    fourth->next = fifth;

    fifth->letter = '!';
    //NULL

    printList(head);

    char str[15]= "Hello, World!";
    //char str[15]= "Wuddup";
    //char *p = str;
    struct Node* second_head = (struct Node*)malloc(sizeof(struct Node)); 
    second_head->letter=NULL;
    second_head->next=NULL;
    second_head->last=NULL;
    makeList(str, second_head);
    printList(second_head);

    return 0;
}

File: doot_example/python3/doot.py
#! /usr/bin/python3
import argparse

class Node:

    def __init__(self, letter):
        self.letter = letter
        self.next = None

class LinkedList:

    def __init__(self):
        self.head = None

    def printList(self):
        tmp = self.head
        while(tmp):
            print (tmp.letter,end = "")
            tmp = tmp.next
    
    def append(self, letter):
        n = Node(letter)
        if self.head is None:
            self.head = n
            return
        last = self.head
        #walk list until end is found, O(n)
        while (last.next):
            last = last.next
        #stitch new node in!
        last.next = n



def main(args):
    if(args.string is not None): print(args.string)
    else:
        print("Doot Doot!")
    
    print("Now using pre-populated Linked List:")
    llist = LinkedList()
    llist.head = Node('D')
    second = Node('o')
    third = Node('o')
    fourth = Node('t')
    fifth = Node('!')
    llist.head.next = second
    second.next = third
    third.next = fourth
    fourth.next = fifth

    llist.printList()
    print("") #newLine hack
    
    if(args.string is not None): 
        print("Now printing provided text with linked list: ")
        llist_second = LinkedList()
        for i,s in enumerate(args.string):
            llist_second.append(args.string[i])
        llist_second.printList()
        print("") #newLine hack

if __name__ == "__main__":
    my_parser = argparse.ArgumentParser()
    my_parser.add_argument('-s','--string',type=str,help="String to print")
    args = my_parser.parse_args()
    main(args)